### PR TITLE
Fix for issue 19983 Can't upload customer Image attribute programmatically

### PR DESCRIPTION
**Pull request by Nazar65**
### Description (*)
Could not load image for customer attribute, error: Base64 is not defined "

### Fixed Issues (if relevant)
1. magento/magento2#19983: Could not load image for customer attribute,

### Manual testing scenarios (*)
1-Create customer attribute type file -> 
```
$eavSetup = $this->eavSetupFactory->create(['setup' => $setup]);
        $eavSetup->addAttribute(
            \Magento\Customer\Model\Customer::ENTITY,
            'sample_attribute2',
            [
                'type'         => 'text',
                'label'        => 'Sample Attribute',
                'input'        => 'file',
                'required'     => false,
                'visible'      => true,
                'user_defined' => false,
                'position'     => 999,
                'system'       => 0,
            ]
        );
        $sampleAttribute = $this->eavConfig->getAttribute(Customer::ENTITY, 'sample_attribute');

        // more used_in_forms ['adminhtml_checkout','adminhtml_customer','adminhtml_customer_address','customer_account_edit','customer_address_edit','customer_register_address']
        $sampleAttribute->setData(
            'used_in_forms',
            ['adminhtml_customer','customer_register_address','customer_address_edit','customer_account_edit']

        );
        $sampleAttribute->save();
```
**2. try to upload image on adminhtml customer edit.**

### Expected Result 
No errors in Console, Image uploaded 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @Nazar65. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Eav/Model/Attribute/Data/File.php [hunk 1]
```diff
@@ -146,7 +146,7 @@ protected function _validateByRules($value)
             return $this->_fileValidator->getMessages();
         }
 
-        if (!empty($value['tmp_name']) && !is_uploaded_file($value['tmp_name'])) {
+        if (empty($value['tmp_name']) && !is_uploaded_file($value['tmp_name'])) {
```
Comment (orlangur) [inline app/code/Magento/Eav/Model/Attribute/Data/File.php]: Will this ever evaluate to true?

File: app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js [hunk 2]
```diff
@@ -16,7 +16,8 @@ define([
     'Magento_Ui/js/form/element/abstract',
     'mage/backend/notification',
     'mage/translate',
-    'jquery/file-uploader'
+    'jquery/file-uploader',
+    'mage/adminhtml/tools'
```
Comment (orlangur) [inline app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js:20]: Why need tools?

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js (hunk 2 above)
Comment (Nazar65) [reply app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js:20]: "Base64 is not defined" when try upload image -> caused by line 173 ->
```
 if (!file.id && file.name) {
                file.id = Base64.mageEncode(file.name);
            }
```

File: app/code/Magento/Eav/Model/Attribute/Data/File.php (hunk 1 above)
Comment (Nazar65) [reply app/code/Magento/Eav/Model/Attribute/Data/File.php]: Yes on event where empty value of ['tmp_name'] and file not  uploaded

File: app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js (hunk 2 above)
Comment (Nazar65) [reply app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js:20]: This is core issue for ->  https://github.com/magento/magento2/issues/18688

**Comment by Nazar65** [comment]
Hi @orlangur I was leave a comment

**Comment by Nazar65** [comment]
@orlangur any updates ?

**Comment by sivaschenko** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, here is your [new Magento instance](https://pr-19988.instances.magento-community.engineering).
Admin access: [https://pr-19988.instances.magento-community.engineering/admin](https://pr-19988.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

File: app/code/Magento/Eav/Model/Attribute/Data/File.php (hunk 1 above)
Comment (sivaschenko) [reply app/code/Magento/Eav/Model/Attribute/Data/File.php]: The condition looks correct to me. If you really need a check for `empty($value['tmp_name'])` - make it a separate condition

Comment (sivaschenko) [reply app/code/Magento/Eav/Model/Attribute/Data/File.php]: However, I'd leave php side as it is

File: app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js (hunk 2 above)
Comment (sivaschenko) [reply app/code/Magento/Ui/view/base/web/js/form/element/file-uploader.js:20]: That's interesting, there are more js files using `Base64` without dependency on `mage/adminhtml/tools` - probably those files require to be fixed too

**Review by sivaschenko — CHANGES_REQUESTED** [review]

**Comment by sivaschenko** [comment]
Hi @Nazar65 thanks for the contribution please see my code review notes

**Comment by Nazar65** [comment]
@sivaschenko Well, I will provide a better fix for this problem, thank you

**Comment by Nazar65** [comment]
@sivaschenko Base64 defined by adminhtml tools please loook -> https://github.com/magento/magento2/blob/9c36f6f237a3b83d6acd8b87c979e73fd8ffbfea/lib/web/mage/adminhtml/grid.js#L6 
There is no difference, all other files are using mage/adminhtml

**Comment by sivaschenko** [comment]
@Nazar65 you are completely right about Base64, I just mentioned that it appears there may be more bugs in Magento because of Base64 usage without dependency on `mage/adminhtml/tools` in several other js files

**Comment by Nazar65** [comment]
@sivaschenko i'm added adminhtml/tools where is used but not defined, what about php part ? for me this is the good condition on "if" statement.

File: app/code/Magento/Eav/Model/Attribute/Data/File.php (hunk 1 above)
Comment (sivaschenko) [inline app/code/Magento/Eav/Model/Attribute/Data/File.php]: Hi @Nazar65 the condition `empty($value['tmp_name']) && !is_uploaded_file($value['tmp_name'])` will evaluate to:
 - `false` is  `$value['tmp_name']` is set
 - `true`if `$value['tmp_name']` is empty

The second part of the condition will be checked only if `$value['tmp_name']` is empty, that does not make sense in this case.

That's why I suggest changing this condition to fail validation in both cases if `tmp_name` is empty or file is not uploaded
```suggestion
        if (empty($value['tmp_name']) || !is_uploaded_file($value['tmp_name'])) {
```

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @Nazar65 thanks for your updates, please see my comment with the suggestion regarding the php condition

File: app/code/Magento/Eav/Model/Attribute/Data/File.php (hunk 1 above)
Comment (Nazar65) [reply app/code/Magento/Eav/Model/Attribute/Data/File.php]: Ohh. :+1:  thanks, now clearly.

**Comment by Nazar65** [comment]
@sivaschenko I'm found the core problem of issue, the function  is_uploaded_file return false because -> `is_uploaded_file — Tells whether the file was uploaded via HTTP POST` But the variable $_FILES['tmp_name'] on validate are empty, because file already downloaded, so i was suggest change this to` !file_exists($value['tmp_name'])`

**Comment by sivaschenko** [comment]
@Nazar65 I am not sure there is a difference between `file_exists` and  `is_uploaded_file` in case when `$value['tmp_name']`.   `is_uploaded_file` should work fine in case when `tmp_name` is set, so I would keep `is_uploaded_file`

**Comment by Nazar65** [comment]
@sivaschenko yes, but  function is_uploaded_file return false but file exist. this function check if file uploaded via HTTP POST action `returns TRUE if the file named by filename was uploaded via HTTP POST` but not checks if file exist.

**Comment by Nazar65** [comment]
@sivaschenko 
is_uploaded
selection_276

file_exists
selection_277

**Comment by Nazar65** [comment]
@sivaschenko so basically to fix this issue we need just add the following condition
```
if (!empty($value['tmp_name']) && !file_exists($value['tmp_name'])) {
            return [__('"%1" is not a valid file.', $label)];
        }
```

**Comment by Nazar65** [comment]
@orlangur or @sivaschenko  any update on this ?

**Comment by Zyles** [comment]
What is the fix? I can't do any development because I can't change themes.

**Review by sivaschenko — APPROVED** [review]

**Comment by sivaschenko** [comment]
Sorry for delay @Nazar65 the pull request is approved and will be further processed internally

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4152](https://jira.corp.magento.com/browse/ENGCOM-4152) has been created to process this Pull Request

**Comment by sivaschenko** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko. Thank you for your request. I'm working on Magento instance for you

**Pull request merged**